### ISSUE TITLE
Illegal application:  The term "do_foo_ind_1" ... cannot be applied to the terms

### ISSUE THREAD

**Issue by ANogin** [mattam82/Coq-Equations#108]
With the latest v8.8, the following code:
```
From Coq.Lists Require Import List.
From Equations Require Import Equations.

(* This type is from VST: 
 * https://github.com/PrincetonUniversity/VST/blob/v2.1/floyd/compact_prod_sum.v#L13 *)
Fixpoint compact_sum (T: list Type): Type :=
  match T with
  | nil => unit
  | t :: nil => t
  | t :: T0 => (t + compact_sum T0)%type
end.

Inductive foo :=
| Sum : list foo -> foo.

Equations foo_type (t : foo) : Type :=
  foo_type (Sum u) := compact_sum (List.map foo_type u).

(* Moving val into the return type, rather than having it as an argument might be unnecessary if
   https://github.com/mattam82/Coq-Equations/issues/73 was fixed *)
Equations (struct t) do_foo (t : foo): forall (val : foo_type t), nat := {
  do_foo (Sum u) := fun val => do_foo_sum u val}

  where (struct fs)
    do_foo_sum (fs : list foo) (val : compact_sum (List.map foo_type fs)) : nat := {
    do_foo_sum nil _ := 0;
    (* Attempting to work around https://github.com/mattam82/Coq-Equations/issues/78 *)
    do_foo_sum (cons hd tl) _ <= (do_foo_sum tl) => {
      do_foo_sum (cons var nil) val _ :=
        do_foo var val;
      do_foo_sum (cons hd tl) val do_foo_tl :=
        match val with
        | inl v => do_foo hd v
        | inr vs => do_foo_tl vs
        end }}.
```
results in
```
Illegal application: 
The term "do_foo_ind_1" of type "forall fs : list foo, compact_sum (map foo_type fs) -> nat -> Prop"
cannot be applied to the terms
 "l" : "list foo"
 "do_foo_sum l" : "compact_sum (map foo_type l) -> nat"
The 2nd term has type "compact_sum (map foo_type l) -> nat" which should be coercible to
 "compact_sum (map foo_type l)".
```

**Comment by ANogin** [comment]
Ah yes, indeed.

**Comment by mattam82** [comment]
Closed via issue #114
